### PR TITLE
feat: do not transmit nil vote

### PR DESCRIPTION
**Pull request by KaoImin**
This PR does: 
* do not transmit nil vote
* use last precommit QC instead of last commit proposal
* fix some filter condition in state machine
* refactor check whether has aggregated votes at the current epoch ID and round
* refactor receive verify epoch response
* remove some useless code

### CODE REVIEW

File: src/state/process.rs [hunk 1]
```diff
@@ -44,23 +43,22 @@ enum MsgType {
...
+    is_full_transcation: HashMap<Hash, bool>,
+    is_leader:           bool,
+    leader_address:      Address,
+    last_commit_qc:      Option<AggregatedVote>,
+    epoch_start:         Instant,
+    epoch_interval:      u64,
+
+    resp_tx:  UnboundedSender<VerifyResp<S>>,
     function: Arc<F>,
     pin_txs:  PhantomData<S>,
```
Comment (rev-chaos) [inline src/state/process.rs]: This PhantomData can be removed

File: src/state/process.rs [hunk 2]
```diff
@@ -203,6 +203,35 @@ where
...
+        let resp = msg.ok_or_else(|| ConsensusError::Other("Event sender dropped".to_string()))?;
+        if resp.epoch_id != self.epoch_id {
+            return Ok(());
+        }
+
+        let epoch_hash = resp.epoch_hash.clone();
+        self.is_full_transcation
+            .insert(epoch_hash, resp.full_txs.is_some());
+
+        if let Ok(qc) = self
```
Comment (rev-chaos) [inline src/state/process.rs]: should first check precommitQC, if exist, trigger commit, else check prevoteQC.

Comment (rev-chaos) [reply src/state/process.rs]: and should search qc by epoch_hash, not by self.epoch_id and self.round.

Comment (rev-chaos) [reply src/state/process.rs]: In fact, the process of check prevoteQC is a little complex. You should first search qc by epoch_hash, and then compare the output qcs' max round, if qcs' max round >= self.round, trigger PrevoteQC.

File: src/state/process.rs [hunk 3]
```diff
@@ -710,6 +690,11 @@ where
         )?;
         self.verify_address(&vote.voter, true)?;
 
+        if epoch_id == self.epoch_id {
```
Comment (rev-chaos) [inline src/state/process.rs]:
```suggestion
        if epoch_id == self.epoch_id - 1 {
```

File: src/state/process.rs [hunk 4]
```diff
@@ -710,6 +690,11 @@ where
         )?;
         self.verify_address(&vote.voter, true)?;
```
Comment (rev-chaos) [inline src/state/process.rs]: lack a critical check, you should check whether self.address is the correct leader of vote's epoch_id and round.

Comment (rev-chaos) [reply src/state/process.rs]: and so in handle_signed_proposal

File: src/state/process.rs [hunk 5]
```diff
@@ -735,8 +720,15 @@ where
             return Ok(());
```
Comment (rev-chaos) [inline src/state/process.rs]: remove if !self.is_leader filter condition. A node should process any vote which pass above filter condition by its round.

Comment (KaoImin) [reply src/state/process.rs]: It seems useless. If self is a leader, receive a future signed vote is impossible.

File: src/state/process.rs (hunk 4 above)
Comment (KaoImin) [reply src/state/process.rs]: This is handling signed vote, not handling aggregated vote. The critical check has occurred while handling signed proposal.

File: src/state/process.rs [hunk 6]
```diff
@@ -696,8 +678,6 @@ where
...
         }
 
-        // TODO: handle signed vote with epoch ID == self.epoch_id - 1
```
Comment (rev-chaos) [inline src/state/process.rs]: simplify filter condition

File: src/state/process.rs (hunk 4 above)
Comment (rev-chaos) [reply src/state/process.rs]: yes, you are right.

File: src/state/process.rs (hunk 5 above)
Comment (rev-chaos) [reply src/state/process.rs]: yes, ignore it

File: src/state/process.rs [hunk 7]
```diff
@@ -823,22 +817,31 @@ where
...
+            Ordering::Less => {
+                debug!(
+                    "Overlord: state receive an outdated QC, epoch ID {}, round {}",
+                    epoch_id, round,
+                );
+                return Ok(());
+            }
+
+            Ordering::Greater => {
+                if self.epoch_id + FUTURE_EPOCH_GAP > epoch_id {
```
Comment (rev-chaos) [inline src/state/process.rs]: miss round gap filter

File: src/state/process.rs [hunk 8]
```diff
@@ -849,7 +852,10 @@ where
             qc_type.clone(),
         )?;
 
-        if epoch_id == self.epoch_id && round > self.round && self.round + FUTURE_ROUND_GAP > round
+        if qc_type == VoteType::Prevote
+            && epoch_id == self.epoch_id
+            && round > self.round
+            && self.round + FUTURE_ROUND_GAP > round
```
Comment (rev-chaos) [inline src/state/process.rs]: do not need round gap filter here, because in the same epoch_id, no one can mock a valid aggregate vote.

Comment (rev-chaos) [reply src/state/process.rs]: should handle aggregate vote which round > self.round

File: src/state/process.rs [hunk 9]
```diff
@@ -676,28 +658,15 @@ where
...
-            return Ok(());
-        } else if self.epoch_id + FUTURE_EPOCH_GAP < epoch_id {
-            warn!("Overlord: state receive a much higher epoch's vote.");
-            return Ok(());
-        } else if (epoch_id == self.epoch_id && self.round + FUTURE_ROUND_GAP < round)
-            || (epoch_id > self.epoch_id && round > FUTURE_ROUND_GAP)
+        if epoch_id != self.epoch_id - 1
+            || !self.is_leader
+            || epoch_id != self.epoch_id
+            || round != self.round
```
Comment (rev-chaos) [inline src/state/process.rs]: wrong logic

File: src/state/process.rs [hunk 10]
```diff
@@ -849,55 +825,19 @@ where
...
-        debug!("Overlord: state check if get full transcations");
-
-        let mut epoch_hash = qc_hash;
-        if !epoch_hash.is_empty() {
-            if qc_type == VoteType::Prevote {
-                epoch_hash = self.check_full_txs(epoch_hash).await?;
-            } else if !self.try_get_full_txs(&epoch_hash) {
-                return Ok(());
-            }
+        if round != self.round {
```
Comment (rev-chaos) [inline src/state/process.rs]:
```suggestion
        if round > self.round {
```

**Pull request merged**